### PR TITLE
DPDK: add support for add_on_miss table property

### PR DESCRIPTION
**Pull request by hanw**
This PR implements the support for add_on_miss table property in pna. 

A P4 table with add_on_miss property enabled is translated into the learner table object in DPDK runtime.

The removeComplexExpressions pass is refactored into a midend pass from backend/bmv2 to be used in dpdk backend.

### CODE REVIEW

File: backends/dpdk/dpdkProgram.cpp [hunk 1]
```diff
@@ -540,28 +574,34 @@ boost::optional<int> ConvertToDpdkControl::getNumberFromProperty(const IR::P4Tab
...
 
-            selectors.push_back(selector);
-        } else {
-            auto table = new IR::DpdkTable(t->name.toString(), t->getKey(), t->getActionList(),
-                    t->getDefaultAction(), t->properties);
-            tables.push_back(table);
-        }
+        selectors.push_back(selector);
+    } else if (structure->learner_tables.count(t->name.name) != 0) {
+        //
```
Comment (mbudiu-vmw) [inline backends/dpdk/dpdkProgram.cpp:597]: empty comment

File: backends/dpdk/spec.cpp [hunk 2]
```diff
@@ -430,6 +435,45 @@ std::ostream &IR::DpdkSelector::toSpec(std::ostream &out) const {
...
+        0) {
+        out << " args none ";
+    } else {
+        BUG("non-zero default action arguments not supported yet");
+    }
+    out << std::endl;
+    if (auto size = properties->getProperty("size")) {
+        out << "\tsize " << DPDK::toStr(size->value) << "" << std::endl;
+    } else {
+        out << "\tsize 0x10000" << std::endl;
```
Comment (mbudiu-vmw) [inline backends/dpdk/spec.cpp:464]: wouldn't it be nicer to have this be a static variable or something similar?

File: backends/dpdk/spec.cpp [hunk 3]
```diff
@@ -430,6 +435,45 @@ std::ostream &IR::DpdkSelector::toSpec(std::ostream &out) const {
...
+    }
+    if (auto size = properties->getProperty("psa_idle_timeout")) {
+        out << "\ttimeout " << DPDK::toStr(size->value) << "" << std::endl;
+    } else {
+        out << "\ttimeout 120"<< std::endl;
```
Comment (mbudiu-vmw) [inline backends/dpdk/spec.cpp:469]: same for default timeout.

File: testdata/p4_16_samples/pna-add-on-miss.p4 [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    inout main_metadata_t user_meta,     // from main parser, to "next block"
+    in    pna_main_input_metadata_t  istd,
+    inout pna_main_output_metadata_t ostd)
+{
+    action next_hop(PortId_t vport) {
+        send_to_port(vport);
+    }
+    action add_on_miss_action() {
+        bit<32> tmp;
+        add_entry(action_name="next_hop", action_params = tmp);
```
[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply testdata/p4_16_samples/pna-add-on-miss.p4:138]: tmp is not initialized here. What does this mean?

**Review by mbudiu-vmw — CHANGES_REQUESTED** [review]
You are missing the reference outputs.
You need to document the `add_entry` function. The way it is used in this example makes no sense to me.

File: testdata/p4_16_samples/pna-add-on-miss.p4 (hunk 4 above)
[the remark this answers is not in the document]
Comment (hanw) [reply testdata/p4_16_samples/pna-add-on-miss.p4:138]: add_entry is defined in pna.p4 https://github.com/p4lang/pna/blob/main/pna.p4#L668

[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply testdata/p4_16_samples/pna-add-on-miss.p4:138]: It may be defined there, but it is not documented at all.

**Comment by hanw** [comment]
Thanks for the review. I have addressed most of the comments.

**Comment by mbudiu-vmw** [comment]
This seems to be failing some tests. Please ping me again if you want another review.

**Comment by hanw** [comment]
@mbudiu-vmw Please take another look?

File: backends/dpdk/dpdkArch.cpp [hunk 5]
```diff
@@ -1458,5 +1459,70 @@ const IR::Node* SplitP4TableCommon::postorder(IR::SwitchStatement* statement) {
...
+            return;
+        } else {
+            use_add_on_miss = expr->to<IR::BoolLiteral>()->value;
+            if (use_add_on_miss)
+                structure->learner_tables.insert(t->name.name);
+        }
+    }
+
+    /* sanity checks */
+    auto default_action = t->properties->getProperty("default_action");
```
[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply backends/dpdk/dpdkArch.cpp:1480]: So it is ok to have add_on_miss but a default_action which can be changed dynamically by the control plane?

File: backends/dpdk/spec.h [hunk 6]
```diff
@@ -0,0 +1,21 @@
+/*
```
Comment (mbudiu-vmw) [inline backends/dpdk/spec.h:1]: what's this file for?

**Review by mbudiu-vmw — APPROVED** [review]

**Pull request merged**